File: java/09_logging_failures.java
// A09 - Logging Failures. VULN: sensitive data in logs, no alerting.
import java.util.logging.*;

public class LoggingFailures {
  private static final Logger LOG = Logger.getLogger("auth");

  public static void onLogin(String user, String password, boolean ok) {
    // VULN: A09 - credentials in log; no security event on failure
    LOG.info("login user=" + user + " pass=" + password + " ok=" + ok);
  }
}

File: c/09_logging_failures.c
/* A09 - Logging Failures. VULN: sensitive data to syslog, no rotation. */
#include <stdio.h>

void log_secret(const char *api_key) {
  /* VULN: A09 - secrets in logs */
  printf("[INFO] using api_key=%s\n", api_key);
}
